File: WitchcraftApp/src/main/java/com/tarbadev/witchcraft/carts/persistence/entity/CartEntity.java
package com.tarbadev.witchcraft.carts.persistence.entity;

import com.tarbadev.witchcraft.carts.domain.entity.Cart;
import com.tarbadev.witchcraft.recipes.persistence.entity.RecipeEntity;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;

import javax.persistence.*;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Builder
@Data
@NoArgsConstructor
@AllArgsConstructor
@Entity(name = "carts")
public class CartEntity {
  @Id
  @GeneratedValue(strategy = GenerationType.AUTO)
  private Integer id;
  @CreationTimestamp
  protected LocalDateTime createdAt;
  @ManyToMany(cascade = CascadeType.MERGE)
  private List<RecipeEntity> recipes;
  @OneToMany(cascade = CascadeType.ALL)
  @JoinColumn(name = "cart_id")
  private List<ItemEntity> items;

  public CartEntity(Cart cart) {
    id = cart.getId();
        recipes = cart.getRecipes().stream().map(RecipeEntity::new).collect(Collectors.toList());
    items = cart.getItems().stream().map(ItemEntity::new).collect(Collectors.toList());
  }

  public Cart cart() {
    return Cart.builder()
        .id(id)
        .createdAt(createdAt)
        .recipes(recipes.stream().map(RecipeEntity::recipe).collect(Collectors.toList()))
        .items(items.stream().map(ItemEntity::item).collect(Collectors.toList()))
        .build();
  }
}

File: WitchcraftApp/src/main/kotlin/com/tarbadev/witchcraft/carts/persistence/entity/CartEntity.kt
package com.tarbadev.witchcraft.carts.persistence.entity

import com.tarbadev.witchcraft.carts.domain.entity.Cart
import com.tarbadev.witchcraft.recipes.persistence.entity.RecipeEntity
import org.hibernate.annotations.CreationTimestamp
import java.time.Instant
import javax.persistence.*

@Entity(name = "cart")
data class CartEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Int = 0,

    @CreationTimestamp
    var createdAt: Instant = Instant.now(),

    @ManyToMany(cascade = [CascadeType.MERGE])
    @JoinTable(
        name = "cart_recipe",
        joinColumns = [JoinColumn(name = "cart_id")],
        inverseJoinColumns = [JoinColumn(name = "recipe_id")]
    )
    val recipes: Set<RecipeEntity> = emptySet(),

    @OneToMany(cascade = [CascadeType.ALL])
    @JoinColumn(name = "cart_id")
    @OrderBy("name ASC")
    val items: List<ItemEntity> = emptyList()
) {
    companion object {
        fun fromCart(cart: Cart): CartEntity {
            return CartEntity(
                cart.id,
                cart.createdAt,
                cart.recipes.map { RecipeEntity.fromRecipe(it) }.toSet(),
                cart.items.map { ItemEntity.fromItem(it) }
            )
        }
    }

    fun toCart(): Cart {
        return Cart(
            id = id,
            createdAtDateTime = createdAt,
            recipes = recipes.map { it.toRecipe() },
            items = items.map { it.toItem() }
        )
    }
}
